pcdsdevices/lodcm.py

```
@@ -85,42 +853,131 @@ class LODCM(BaseInterface, Device):
+        self.x2 = self.tower2.x2
+        self.y2 = self.tower2.y2
+        self.th2 = self.tower2.th2
+        self.chi2 = self.tower2.chi2
+        self.h2n = self.tower2.h2n
+        self.diode2 = self.tower2.diode2
+        # diagnostic tower
+        self.dh = self.diag_tower.dh
+        self.dv = self.diag_tower.dv
+        self.dr = self.diag_tower.dr
+        self.df = self.diag_tower.df
+        self.dd = self.diag_tower.dd
+        self.yag_zoom = self.diag_tower.yag_zoom
+        # states
+        self.h1n_state = self.tower1.h1n_state
+        self.y1_state = self.tower1.y1_state
+        self.chi1_state = self.tower1.chi1_state
+        self.h2n_state = self.tower2.h2n_state
+        self.y2_state = self.tower2.y2_state
+        self.chi2_state = self.tower2.chi2_state
+        # # offset positioners - tower 1
+        self.th1_si = self.energy_si.th1
+        self.z1_si = self.energy_si.z1
+        self.th1_c = self.energy_c.th1
+        self.z1_c = self.energy_c.z1
+
+        self.x1_c = self.tower1.x1_c
+        self.x1_si = self.tower1.x1_si
+        self.y1_c = self.tower1.y1_c
+        self.y1_si = self.tower1.y1_si
+        self.chi1_c = self.tower1.chi1_c
+        self.chi1_si = self.tower1.chi1_si
+        self.h1n_c = self.tower1.h1n_c
+        self.h1n_si = self.tower1.h1n_si
+        self.h1p_c = self.tower1.h1p_c
+        self.h1p_si = self.tower1.h1p_si
+        # # offset positioners - tower 2
+        self.th2_si = self.energy_si.th2
+        self.z2_si = self.energy_si.z2
+        self.th2_c = self.energy_c.th2
+        self.z2_c = self.energy_c.z2
+
+        self.x2_c = self.tower2.x2_c
+        self.x2_si = self.tower2.x2_si
+        self.y2_c = self.tower2.y2_c
+        self.y2_si = self.tower2.y2_si
+        self.chi2_c = self.tower2.chi2_c
+        self.chi2_si = self.tower2.chi2_si
+        self.h2n_c = self.tower2.h2n_c
+        self.h2n_si = self.tower2.h2n_si
+
+    @property
+    def energy(self):
```

I think I should not have this property here like that.... might not be good practice? since it is an actual object component not a simple property...

The idea here is that I wanted the user to be able to do something like this:
```
lom.energy.move(10)
```
instead of 
```
lom.energy_si.move(10) or lom.energy_c.move(10)
```
but there's probably a better way of dealing with this

---

I think this is a reasonable solution to the problem, the only other one I can think of is an "energy" motor object that picks whether to forward the commands to instead of a property that returns one or the other. But I'm not sure which approach is better.

Either way, the only thing that seems "wrong" here is returning `None` sometimes (and maybe having properly access wait for EPICS, that's usually not great too)

---

Ok I can probably raise an error here instead of returning `None`... ? 
Can you please elaborate on this: ***(and maybe having properly access wait for EPICS, that's usually not great too)***

---

I think raising an exception or just picking one to use as the default are both fine.

When you access a python property, there is some assumption that it will be "fast" (e.g., just checking a value)

In this case, accessing `lodcm.energy` can be slow if EPICS is being slow, since you have to check the material type before deciding which underlying object to return.

---

Ohhh got it, thank you. 

I'm just going to pick one as default I think for now - but this won't solve the wait access on the property - so maybe your previous suggestion with an `energy` proxy motor object might be better, i don't know for sure how it would look like yet

---

It's probably correct to not worry about that so much for now and consider changing it in the future if it becomes a problem. A lot of times these sorts of things never actually become a problem.

---

Ok sounds good, pasting here for future reference - I think it will look similar to this example here: https://ruddra.com/python-proxy-object/